plugins/inputs/prometheus/kubernetes.go

```diff
@@ -134,76 +133,43 @@ func (p *Prometheus) watchPod(ctx context.Context, clientset *kubernetes.Clients
 	podinformer := informerfactory.Core().V1().Pods()
 	podinformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(newObj interface{}) {
-			key, err := cache.MetaNamespaceKeyFunc(newObj)
-			if err != nil {
-				p.Log.Errorf("getting key from cache %s\n", err.Error())
-			}
-
-			namespace, name, err := cache.SplitMetaNamespaceKey(key)
-			if err != nil {
-				p.Log.Errorf("splitting key into namespace and name %s\n", err.Error())
+			newPod, ok := newObj.(*corev1.Pod)
+			if !ok {
+				p.Log.Error("[BUG] Not a Pod, report it to the Github issues, please")
+				return
 			}
-
-			pod, _ := clientset.CoreV1().Pods(namespace).Get(ctx, name, metav1.GetOptions{})
-
-			if shouldScrapePod(pod, p) {
-				registerPod(pod, p)
+			if shouldScrapePod(newPod, p) {
+				registerPod(newPod, p)
 			}
 		},
-		UpdateFunc: func(oldObj, newObj interface{}) {
-			newKey, err := cache.MetaNamespaceKeyFunc(newObj)
-			if err != nil {
-				p.Log.Errorf("getting key from cache %s\n", err.Error())
+		// On Pod status updates and regular reList by Informer
+		UpdateFunc: func(_, newObj interface{}) {
+			newPod, ok := newObj.(*corev1.Pod)
+			if !ok {
+				p.Log.Error("[BUG] Not a Pod, report it to the Github issues, please")
+				return
 			}
 
-			newNamespace, newName, err := cache.SplitMetaNamespaceKey(newKey)
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(newObj)
 			if err != nil {
-				p.Log.Errorf("splitting key into namespace and name %s\n", err.Error())
+				p.Log.Errorf("getting key from cache %s", err.Error())
 			}
-
-			newPod, _ := clientset.CoreV1().Pods(newNamespace).Get(ctx, newName, metav1.GetOptions{})
-
+			podID := PodID(key)
 			if shouldScrapePod(newPod, p) {
-				if newPod.GetDeletionTimestamp() == nil {
+				// When Informers re-Lists, pod might already be registered,
+				// do nothing if it is, register otherwise
+				if _, ok = p.kubernetesPods[podID]; !ok {
 					registerPod(newPod, p)
 				}
-			}
-
-			oldKey, err := cache.MetaNamespaceKeyFunc(oldObj)
-			if err != nil {
-				p.Log.Errorf("getting key from cache %s\n", err.Error())
-			}
-
-			oldNamespace, oldName, err := cache.SplitMetaNamespaceKey(oldKey)
-			if err != nil {
-				p.Log.Errorf("splitting key into namespace and name %s\n", err.Error())
-			}
-
-			oldPod, _ := clientset.CoreV1().Pods(oldNamespace).Get(ctx, oldName, metav1.GetOptions{})
-
-			if shouldScrapePod(oldPod, p) {
-				if oldPod.GetDeletionTimestamp() != nil {
-					unregisterPod(oldPod, p)
```

Comment:
Any checks to `GetDeletionTimestamp` are now gone, why?

Reply:
we don't need it, we base our actions only on the new state of the pod, regardless what was the previous state. Previously code accounted for pod name change, no update to the pod can change it's name, only delete+create can.